File: test/files/password-generator.cs
// C# with Elan 2.0.0-beta

// Ideas for further development

// - rule for no more than 'n' same characters in succession

const String upperChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

const String lowerChars = "abcdefghijklmnopqrstuvwxyz";

const String digitChars = "01234567890";

const String symbolChars = "!$%^&*";

const Int length = 32;

const Boolean upper = true;

const Boolean lower = true;

const Boolean digit = true;

const Boolean symbol = true;

static void main() {
  var all = upperChars + lowerChars + digitChars + symbolChars;
  var password = "";
  var valid = false;
  while (!valid) {
    var pwRef = new AsRef<string>("");
    populatePassword(pwRef, all); // call procedure
    password = pwRef.value(); // reassign variable
    valid = isValid(password, upper, lower, digit, symbol); // reassign variable
  } // end while
  Console.WriteLine(password); // print
} // end main

static void populatePassword(AsRef<string> password, string all) { // procedure
  foreach (var i in range(1, length + 1)) {
    var rnd = randint(0, all.length() - 1);
    password.set(password.value() + all[rnd]); // call procedure
  } // end foreach
} // end procedure

static bool isValid(string password, bool mustHaveUpper, bool mustHaveLower, bool mustHaveDigit, bool mustHaveSymbol) { // function
  var validForUpper = passesRule(mustHaveUpper, upperChars, password);
  var validForLower = passesRule(mustHaveLower, lowerChars, password);
  var validForDigits = passesRule(mustHaveDigit, digitChars, password);
  var validForSymbols = passesRule(mustHaveSymbol, symbolChars, password);
  return validForUpper && validForLower && validForDigits && validForSymbols;
} // end function

[TestMethod] static void test_isValid() {
  Assert.AreEqual(true, isValid("$4De", true, true, true, true));
  Assert.AreEqual(false, isValid("$4de", true, true, true, true));
  Assert.AreEqual(true, isValid("$4de", false, true, true, true));
  Assert.AreEqual(true, isValid("eD$4", true, true, true, true));
  Assert.AreEqual(true, isValid("$4De$4De", true, true, true, true));
} // end test

static bool passesRule(bool rule, string charSet, string password) { // function
  return (!rule) || hasAtLeastOneFrom(charSet, password);
} // end function

[TestMethod] static void test_passesRule() {
  Assert.AreEqual(true, passesRule(true, "12A", "ABC"));
  Assert.AreEqual(true, passesRule(false, "12A", "ABC"));
  Assert.AreEqual(false, passesRule(true, "12", "ABCD"));
  Assert.AreEqual(true, passesRule(false, "12", "ABCD"));
} // end test

static bool hasAtLeastOneFrom(string fromChars, string password) { // function
  var hasOne = false;
  foreach (var ch in password) {
    hasOne = hasOne || fromChars.contains(ch); // reassign variable
  } // end foreach
  return hasOne;
} // end function

[TestMethod] static void test_hasAtLeastOneFrom() {
  Assert.AreEqual(true, hasAtLeastOneFrom("12A", "ABC"));
  Assert.AreEqual(true, hasAtLeastOneFrom("C12", "ABC"));
  Assert.AreEqual(false, hasAtLeastOneFrom("12", "ABCD"));
} // end test

File: test/files/password-generator.vb
' VB.NET with Elan 2.0.0-beta1

' Ideas for further development

' - rule for no more than 'n' same characters in succession

Const upperChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

Const lowerChars = "abcdefghijklmnopqrstuvwxyz"

Const digitChars = "01234567890"

Const symbolChars = "!$%^&*"

Const length = 32

Const upper = True

Const lower = True

Const digit = True

Const symbol = True

Sub main()
  Dim all = upperChars + lowerChars + digitChars + symbolChars ' variable definition
  Dim password = "" ' variable definition
  Dim valid = False ' variable definition
  While Not valid
    Dim pwRef = New AsRef(Of String)("") ' variable definition
    populatePassword(pwRef, all) ' call procedure
    password = pwRef.value() ' reassign variable
    valid = isValid(password, upper, lower, digit, symbol) ' reassign variable
  End While
  Console.WriteLine(password) ' print
End Sub

Sub populatePassword(password As AsRef(Of String), all As String) ' procedure
  For Each i In range(1, length + 1)
    Dim rnd = randint(0, all.length() - 1) ' variable definition
    password.set(password.value() + all(rnd)) ' call procedure
  Next i
End Sub

Function isValid(password As String, mustHaveUpper As Boolean, mustHaveLower As Boolean, mustHaveDigit As Boolean, mustHaveSymbol As Boolean) As Boolean
  Dim validForUpper = passesRule(mustHaveUpper, upperChars, password) ' variable definition
  Dim validForLower = passesRule(mustHaveLower, lowerChars, password) ' variable definition
  Dim validForDigits = passesRule(mustHaveDigit, digitChars, password) ' variable definition
  Dim validForSymbols = passesRule(mustHaveSymbol, symbolChars, password) ' variable definition
  Return validForUpper And validForLower And validForDigits And validForSymbols
End Function

<TestMethod> Sub test_isValid()
  Assert.AreEqual(True, isValid("$4De", True, True, True, True))
  Assert.AreEqual(False, isValid("$4de", True, True, True, True))
  Assert.AreEqual(True, isValid("$4de", False, True, True, True))
  Assert.AreEqual(True, isValid("eD$4", True, True, True, True))
  Assert.AreEqual(True, isValid("$4De$4De", True, True, True, True))
End Sub

Function passesRule(rule As Boolean, charSet As String, password As String) As Boolean
  Return (Not rule) Or hasAtLeastOneFrom(charSet, password)
End Function

<TestMethod> Sub test_passesRule()
  Assert.AreEqual(True, passesRule(True, "12A", "ABC"))
  Assert.AreEqual(True, passesRule(False, "12A", "ABC"))
  Assert.AreEqual(False, passesRule(True, "12", "ABCD"))
  Assert.AreEqual(True, passesRule(False, "12", "ABCD"))
End Sub

Function hasAtLeastOneFrom(fromChars As String, password As String) As Boolean
  Dim hasOne = False ' variable definition
  For Each ch In password
    hasOne = hasOne Or fromChars.contains(ch) ' reassign variable
  Next ch
  Return hasOne
End Function

<TestMethod> Sub test_hasAtLeastOneFrom()
  Assert.AreEqual(True, hasAtLeastOneFrom("12A", "ABC"))
  Assert.AreEqual(True, hasAtLeastOneFrom("C12", "ABC"))
  Assert.AreEqual(False, hasAtLeastOneFrom("12", "ABCD"))
End Sub
